File: brainduck.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <chrono>
#include <cctype>
#include <cstdio>

using namespace std;

void showmem(const vector<unsigned char>& tape) {
	cout << "\nMemory dump:\n";
	
	const int chunk_size = 8;
	
	for (size_t i = 0; i < tape.size(); i += chunk_size) {
		// offset
		printf("%08lx  ", i);
		
		// hex bytes
		for (size_t j = i; j < i + chunk_size && j < tape.size(); j++) {
			printf("%02x ", tape[j]);
		}
		
		// padding for shorter last line
		for (size_t j = tape.size() - i; j < chunk_size; j++) {
			printf("   ");
		}
		
		printf("  ");
		
		// ASCII (safe)
		for (size_t j = i; j < i + chunk_size && j < tape.size(); j++) {
			unsigned char b = tape[j];
			if (isprint(b) || b == ' ') {
				printf("%c", (char)b);
			} else {
				printf(".");
			}
		}
		
		printf("\n");
	}
}

void bf(const string& code) {
	auto start = chrono::high_resolution_clock::now();
	
	// Initialize tape and pointers
	vector<unsigned char> tape(64, 0);
	size_t ptr = 0;
	size_t max_ptr = 0;
	
	// Build jump table
	vector<size_t> jump(code.length(), 0);
	vector<size_t> stack;
	for (size_t i = 0; i < code.length(); i++) {
		if (code[i] == '[') {
			stack.push_back(i);
		}
		if (code[i] == ']') {
			size_t j = stack.back();
			stack.pop_back();
			jump[i] = j;
			jump[j] = i;
		}
	}
	
	// Execute
	size_t ip = 0;
	while (ip < code.length()) {
		switch (code[ip]) {
			case '>':
				ptr++;
				max_ptr = max(max_ptr, ptr);
				// Grow tape if needed
				if (ptr >= tape.size()) {
					tape.resize(ptr + 1, 0);
				}
				break;
			case '<':
				ptr--;
				break;
			case '+':
				tape[ptr]++;
				break;
			case '-':
				tape[ptr]--;
				break;
			case '.':
				putchar(tape[ptr]);
				break;
			case ',':
				tape[ptr] = getchar();
				break;
			case '[':
				if (tape[ptr] == 0) {
					ip = jump[ip];
				}
				break;
			case ']':
				if (tape[ptr] != 0) {
					ip = jump[ip];
				}
				break;
			default:
				break;
		}
		ip++;
	}
	
	auto elapsed = chrono::high_resolution_clock::now() - start;
	
	cout << "\n---\n";
	cout << "execution time = " << chrono::duration_cast<chrono::microseconds>(elapsed).count() << " us\n";
	cout << "max memory usage = " << hex << (max_ptr + 1) << " bytes\n";
	
	showmem(tape);
}

int main(int argc, char* argv[]) {
	if (argc < 2) {
		cerr << "Usage: " << argv[0] << " <file_path>" << endl;
		return 1;
	}
	
	// Read file
	ifstream file(argv[1]);
	if (!file) {
		cerr << "Could not read file" << endl;
		return 1;
	}
	
	string code((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
	file.close();
	
	bf(code);
	
	return 0;
}

File: brainduck.rs
use std::env;
use std::fs;
use std::io::{self, Read};
use std::time::Instant;

fn showmem(mem: Vec<u8>) {
    println!("\nMemory dump:");

    let chunk_size = 8;

    for (i, chunk) in mem.chunks(chunk_size).enumerate() {
        // offset
        print!("{:08x}  ", i * chunk_size);

        // hex bytes
        for b in chunk {
            print!("{:02x} ", b);
        }

        // padding for shorter last line
        for _ in 0..(chunk_size - chunk.len()) {
            print!("   ");
        }

        print!("  ");

        // ASCII (safe)
        for &b in chunk {
            let ch = if b.is_ascii_graphic() || b == b' ' {
                b as char
            } else {
                '.'
            };
            print!("{}", ch);
        }

        println!();
    }
}

fn bf(code: Vec<char>) {
    let start = Instant::now();

    let mut tape = vec![0u8; 64];
    let mut ptr = 0usize;
    let mut max_ptr = 0_usize;

    // build jump table
    let mut stack = vec![];
    let mut jump = vec![0usize; code.len()];
    for (i, &c) in code.iter().enumerate() {
        if c == '[' {
            stack.push(i);
        }
        if c == ']' {
            let j = stack.pop().unwrap();
            jump[i] = j;
            jump[j] = i;
        }
    }

    let mut ip = 0usize;
    while ip < code.len() {
        match code[ip] {
            '>' => {
                ptr += 1;
                max_ptr = max_ptr.max(ptr)
            }
            '<' => ptr -= 1,
            '+' => tape[ptr] = tape[ptr].wrapping_add(1),
            '-' => tape[ptr] = tape[ptr].wrapping_sub(1),
            '.' => print!("{}", tape[ptr] as char),
            ',' => tape[ptr] = io::stdin().bytes().next().unwrap().unwrap(),
            '[' if tape[ptr] == 0 => ip = jump[ip],
            ']' if tape[ptr] != 0 => ip = jump[ip],
            _ => {}
        }
        ip += 1;
    }

    let elapsed = start.elapsed();

    println!("\n---");

    println!("execution time = {:?}", elapsed);
    println!("max memory usage = {:x} bytes", max_ptr + 1);

    showmem(tape);
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 {
        eprintln!("Usage: {} <file_path>", args[0]);
        return;
    }

    let file_path = &args[1];

    let prog: Vec<char> = fs::read_to_string(file_path)
        .expect("Could not read file")
        .chars()
        .collect();

    bf(prog);
}
